include/circt/Dialect/RTG/IR/RTGTypes.td

```diff
@@ -47,4 +47,24 @@ class SetTypeOf<Type elementType> : ContainerType<
   elementType, SetType.predicate, 
   "llvm::cast<rtg::SetType>($_self).getElementType()", "set">;
 
+def DictType : RTGTypeDef<"Dict"> {
+  let summary = "a dictionary";
+  let description = [{
+    This type is a dictionary with a static set of entries. This datatype does
+    not make any assumptions about how the values are stored (could be a struct,
+    a map, etc.). Furthermore, two values of this type should be considered
+    equivalent if they have the same set of entry names and types and the values
+    match for each entry, independent of the order.
+  }];
+
+  let parameters = (ins
+    ArrayRefParameter<"mlir::StringAttr", "dict entry names">:$entryNames,
+    ArrayRefParameter<"mlir::Type", "dict entry types">:$entryTypes);
```

Comment:
you might consider making this a single array of a pair-like data structure.  You wouldn't need the verifier anymore.